cnd/src/network/protocol.rs

```diff
@@ -0,0 +1,207 @@
+use crate::swap_protocols::{
+    rfc003::{Secret, SecretHash},
+    SwapId,
+};
+use futures::{future, AsyncRead, AsyncWrite, Future};
+use libp2p::core::{upgrade, InboundUpgrade, OutboundUpgrade, UpgradeInfo};
+use serde::{Deserialize, Serialize};
+use std::{fmt, io, iter, pin::Pin, str::FromStr};
+use tracing::{debug, trace};
+
+/// The secret hash sharing protocol works in the following way:
+///
+/// - Dialer (Alice) writes the `SecretHashInfo` to the substream.
+/// - Listener (Bob) reads the `SecretHashInfo` from the substream.
+
+/// Represents a prototype for an upgrade to handle the secret hash sharing
+/// protocol.
+#[derive(Debug, Clone)]
+pub struct SecretHashProtocolConfig;
+
+/// Data sent to peer in secret hash protocol.
+#[derive(Deserialize, Debug, Serialize)]
+pub struct SecretHashInfo {
+    swap_id: SwapId,
+    secret_hash: SecretHash,
+}
+
+lazy_static::lazy_static! {
+    static ref JSON_INFO_LENGTH: usize = {
+        let swap_id = SwapId::from_str("7f3a105d-ecf2-4cc6-b35c-b4351ac28a34").unwrap();
+
+        let bytes = b"hello world, you are beautiful!!";
+        let secret = Secret::from(*bytes);
+        let secret_hash = secret.hash();
+
+        let info = SecretHashInfo {
+            swap_id,
+            secret_hash,
+        };
+
+        let serialized = serde_json::to_vec(&info).unwrap();
+
+        serialized.len()
+    };
+}
+
+/// The substream on which info we wish to push is sent.
+pub struct PushSubstream<T> {
+    inner: T,
+}
+
+impl<T> fmt::Debug for PushSubstream<T> {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.debug_tuple("PushSubstream").finish()
+    }
+}
+
+impl<T> PushSubstream<T>
+where
+    T: AsyncWrite + Unpin,
+{
+    /// Sends the information we wish to push on the substream.
+    ///
+    /// Consumes the substream, returning a future that resolves
+    /// when the info has been sent on the underlying connection.
+    pub async fn send(mut self, info: SecretHashInfo) -> Result<(), io::Error> {
+        debug!("Sending secret hash info to peer");
+
+        let message = serde_json::to_vec(&info).unwrap();
+        upgrade::write_one(&mut self.inner, &message).await
+    }
+}
+
+impl UpgradeInfo for SecretHashProtocolConfig {
+    type Info = &'static [u8];
+    type InfoIter = iter::Once<Self::Info>;
+
+    fn protocol_info(&self) -> Self::InfoIter {
+        iter::once(b"/comit/swap/secret_hash/1.0.0")
+    }
+}
+
+impl<C> InboundUpgrade<C> for SecretHashProtocolConfig
+where
+    C: AsyncRead + AsyncWrite + Unpin + Send + Sync + 'static,
+{
+    type Output = SecretHashInfo;
+    type Error = upgrade::ReadOneError;
+    type Future = Pin<Box<dyn Future<Output = Result<Self::Output, Self::Error>> + Send>>;
+
+    fn upgrade_inbound(self, mut socket: C, _: Self::Info) -> Self::Future {
+        trace!("Upgrading inbound connection");
+        Box::pin(async move {
+            let msg = upgrade::read_one(&mut socket, *JSON_INFO_LENGTH).await?;
```

Comment:
Did you do research/testing we actually need to pass the exact same length here? I wouldn't want to have this complexity if this is just the max buffer size and also works if the message is shorter.